### DISCUSSION TITLE
Livewire validateOnly() no longer validates entire array if an element changed

### DISCUSSION THREAD

**Discussion by jelib3an** [livewire/livewire#4675] [7. Bugs (archived)]
Affected versions: ^2.6 (I suspect due to the new nested data binding feature)

If a rule exists for a nested array, it doesn't apply when calling `validateOnly()` on one of it's elements. The impact is, user expects the array to get validated but ends up not, which could be problematic. 

To reproduce, here is an example component with failing test case.

```php
<?php

namespace App\Http\Livewire;

use Illuminate\Support\Facades\Validator;
use Livewire\Component;

class Example extends Component
{
    public $levels = [
        ['id' => 1, 'name' => 'basement'],
    ];

    public $message;

    protected function rules()
    {
        return [
            'levels.*' => function ($attribute, $value, $fail) {
                if (!is_numeric($value['id'])) {
                    $fail('invalid');
                }
            },
        ];
    }

    public function updated($name)
    {
        $this->validateOnly($name);
    }

    public function checkForErrors($name)
    {
        //dd($this->getErrorBag()->all());
        $errors = $this->getErrorBag();
        if ($errors->has($name)) {
            $this->message = collect($errors->get($name))->first();
        }
    }

    public function render()
    {
        return view('livewire.example');
    }
}
```

```php
<?php

namespace Tests\Feature;

use App\Http\Livewire\Example;
use Illuminate\Foundation\Testing\RefreshDatabase;
use Livewire\Livewire;
use Tests\TestCase;

class ExampleTest extends TestCase
{
    /** @test */
    public function livewire_validates_entire_array_if_element_changes()
    {
        Livewire::test(Example::class)
            ->set('levels.0.id', 'not-numeric')
            ->call('checkForErrors', 'levels.0')
            ->assertSet('message', 'invalid');
    }
}
```

Notice, if I set `levels.0.id`, it doesn't run the validation rule for `levels.*`. And when I look for an error message using the key `levels.0`, there is no message. This used to work in 2.5.5. If you downgrade to Livewire 2.5.5, the tests will pass.

I suppose one can argue that the rule for the array should not apply to any individual elements. And the previous logic should have been a bug. But since 2.6 is a minor release, one would not expect any breaking changes. 

As a workaround, I can use Laravel's Validator instead, but it's not as nice because Livewire does a better job of parsing the keys for rules and error messages.

**Comment by jelib3an** [comment]
Actually, found a better workaround. I can just detect if the updated is an array element and just call `validateOnly()` on the entire array

```php
public function updated($name)
{
    $this->validateOnly($name);

    if (preg_match("/^levels\.(\d+)\./", $name, $matches)) {
        $this->validateOnly("levels.{$matches[1]}");
    }
}
```

**Comment by joshhanley** [comment]
@jelib3an yeah I'm not sure if it was a bug that was "fixed" or an issue with a feature, but if it was an issue introduced by a new feature, then PR's are welcome to fix that!

If it's a bug fix though, we try our best not to having breaking changes, but if someone depends on a bug that eventually gets fixed, there's not a lot we can do about that other than try and work around it as best as we can.

**Reply by jelib3an** [reply]
Yeh, the more I think about it, the more it seems the original logic was a bug. Given that there is a workaround and unless more people complain about it, I'd be happy to consider this a non-issue. Thanks!

**Reply by joshhanley** [reply]
@jelib3an yep sounds good to me 🙂
